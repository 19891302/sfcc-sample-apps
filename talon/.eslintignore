**/node_modules/*
**/dist/*
target
coverage
**/__tests__/fixtures/*