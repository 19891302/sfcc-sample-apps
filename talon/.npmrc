# We have two registries we could use:
# Internal (private)
registry=https://nexus.soma.salesforce.com/nexus/content/groups/npm-all/