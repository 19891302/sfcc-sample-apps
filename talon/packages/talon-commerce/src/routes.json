[
    {
        "name": "home",
        "path": "/",
        "isRoot": true,
        "view": "home"
    },
    {
        "name": "search",
        "path": "/search/:query",
        "isRoot": false,
        "view": "search"
    },
    {
        "name": "about",
        "path": "/about",
        "view": "about"
    },
    {
        "name": "error",
        "path": "/error",
        "isDefault": true,
        "view": "error"
    }
]