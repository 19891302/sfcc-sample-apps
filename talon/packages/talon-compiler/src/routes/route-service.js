const { resourceDescriptorToString, RESOURCE_TYPES } = require('talon-common');
const validateRoutes = require('./validate-routes');
const { getView } = require('../metadata/metadata-service');
const TALON_RESOURCES = require('talon-framework/dist/resources.json');
const TALON_RESOURCE_DESCRIPTORS = [...Object.keys(TALON_RESOURCES)];

// TODO remove
const IGNORED_RESOURCES = [
    "view://customPerf"
];

/**
 * Get the resources the given routes depend on.
 *
 * @param  {Object|Object[]} routes the route or routes to get the resources for
 * @returns an array containing the resource descriptors
 *          the given routes depend on
 */
function getRoutesResources(routes, theme, locale) {
    // using `[].concat(routes)` allows the caller to pass either a single route (the current route for example)
    // or an array of routes
    const routesArr = [].concat(routes || []);

    // get theme layouts by type from theme
    const themeLayoutViews = {};

    Object.entries(theme.themeLayouts).forEach(([type, themeLayout]) => {
        themeLayoutViews[type] = themeLayout.view;
    });

    // add the view resources for each route
    const routeViewDescriptors = routesArr.filter(route => route.view).reduce((acc, route) => {
        const results = [];
        results.push(route.view);

        const { themeLayoutType } = getView(route.view);

        if (themeLayoutViews[themeLayoutType]) {
            results.push(themeLayoutViews[themeLayoutType]);
        }
        return acc.concat(results);
    }, []).map(view => {
        return resourceDescriptorToString({ type: RESOURCE_TYPES.VIEW, name: view, locale });
    });

    //
    const resources = [
        ...TALON_RESOURCE_DESCRIPTORS,
        ...routeViewDescriptors
    ];

    // remove duplicates and filter out ignored resources
    return [...new Set(resources)].filter(resourceDescriptor => {
        return !IGNORED_RESOURCES.includes(resourceDescriptor.split('@')[0]);
    });
}

/**
 * Get the route matching the given path.
 *
 * @param {Object[]} routes The routes to get the current route from
 * @param {string} path The path, without basePath
 * @returns {Object} The route matching the given path, or the default route
 */
function getRouteMatchingPath(routes, path) {
    const currentRoute = path && routes.find(route => path === route.path);
    return currentRoute || routes.find(route => route.isDefault);
}

/**
 * Get all views which are routed
 */
function getViewToThemeLayoutMap(routes, theme) {
    const viewToThemeLayoutMap = {};
    routes.forEach((route) => {
        const view = getView(route.view);
        const themeLayout = theme.themeLayouts[view.themeLayoutType];
        viewToThemeLayoutMap[route.view] = themeLayout.view;
    });
    return viewToThemeLayoutMap;
}

module.exports = { getRoutesResources, getRouteMatchingPath, validateRoutes, getViewToThemeLayoutMap };