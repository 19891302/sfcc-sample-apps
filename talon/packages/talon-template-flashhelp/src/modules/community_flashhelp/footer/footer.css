.footer {
    position: relative;
    z-index: 1030;
    width: 100%;
    background-color: var(--lwc-footer-background-color);
    clear: both;
}

.container-fluid {
    padding-top: 10px;
}

