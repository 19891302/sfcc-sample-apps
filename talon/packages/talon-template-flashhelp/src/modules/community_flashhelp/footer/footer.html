<template>
    <footer class="footer">
        <div class="container-fluid slds-text-align_center">
            <p>
                <span><small>Â© 2018 The American National Red Cross</small></span>
            </p>
            <span><small>For more information, please visit redcross.org or cruzrojaamericana.org</small></span>
            <div class="slds-align_absolute-center">
                <ul class="slds-list_horizontal slds-has-inline-block-links_space">
                    <li>
                        <talon-router-link route="home" label="Home"></talon-router-link>
                    </li>
                    <li>
                        <talon-router-link route="about" label="About"></talon-router-link>
                    </li>
                    <li>
                        <talon-router-link route="faq" label="FAQ"></talon-router-link>
                    </li>
                </ul>
            </div>
        </div>
    </footer>    
</template>