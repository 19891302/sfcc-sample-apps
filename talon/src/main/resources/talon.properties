pom.version=${pom.version}
project.artifactId=${project.artifactId}