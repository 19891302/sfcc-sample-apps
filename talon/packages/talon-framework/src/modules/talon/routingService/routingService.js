import { getBasePath } from "talon/configProvider";
import { assert, autoBind, getQueryStringParameterByName } from "talon/utils";
import { getThemeLayoutByView } from "talon/themeService";

const WHITELISTED_PARAMETERS = ['mode'];

/**
 * Routing service class.
 *
 * A single instance of it will be used throughout the app
 * and selected methods will be exported.
 *
 * We still export the class itself for testing purpose so that we can
 * create as many instances as needed.
 */
export class RoutingService {
    router;
    routesByName = {};
    observers = new Set();
    currentRoute;

    /**
     * Register a set of routes and start the router.
     *
     * See JSON schema at https://salesforce.quip.com/EDbbAKBI92ZM#LeeACA0AB7l.
     *
     * See validate-routes.js#validateRoutes() for the validation rules.
     *
     * @param {Object[]} routes - The routes to register
     */
    registerRoutes(routes) {
        assert(this.router, "Router implementation not set");

        routes.forEach((route) => {
            // save route for lookup by name
            this.routesByName[route.name] = route;

            const callback = this.onRouteChange.bind(this, route.name);

            if (route.isDefault) {
                this.router('*', callback);
            } else {
                this.router(route.path, callback);
            }
        });

        // set base path
        const basePath = getBasePath();
        if (basePath) {
            this.router.base(basePath);
        }

        // start the router
        this.router.start();
    }

    /**
     * Get the URL for a given route.
     * This can be used to populate links href.
     *
     * The returned URL will include any whitelisted parameters
     * specified in the current URL.
     *
     * TODO handle route parameters
     *
     * @param {string} name - The route name
     */
    getRouteUrl(name) {
        assert(this.routesByName[name], `Unknown route: ${name}`);

        const routeUrl = getBasePath() + (this.routesByName[name] || {}).path;

        // add present whitelisted parameters to the route URL
        const params = WHITELISTED_PARAMETERS.reduce((acc, param) => {
            const value = getQueryStringParameterByName(param);
            return (value && acc.concat([param + '=' + encodeURIComponent(value)])) || acc;
        }, []);

        if (Object.keys(params).length > 0) {
            return routeUrl + '?' + params.join('&');
        }

        return routeUrl;
    }

    /**
     * Returns a route which has been registered
     *
     * @param {string} name
     */
    getRoute(name) {
        assert(this.routesByName[name], `Unknown route: ${name}`);
        return this.routesByName[name] || {};
    }

    /**
     * Callback invoked by the router when route has changed.
     *
     * This will either reload the whole page if the current and target
     * theme layouts are different, or notify the observers if not.
     *
     * @param {string} name the target route name
     */
    onRouteChange(name) {
        const route = this.routesByName[name];

        // the app container and the router container so that it can create and render
        // the page component and the theme layout
        this.currentRoute = route;
        this.themeLayout = getThemeLayoutByView(route.view);
        this.observers.forEach(observer => {
            observer.next(route);
        });
    }

    /**
     * Navigates to the route with the given name.
     *
     * TODO handle route parameters
     *
     * @param {string} name - The route name
     */
    navigateToRoute(name, params) {
        let path;
        const route = this.getRouteUrl(name).split('/:');

        path = route[0];

        if ( params && route.length > 0 ) {
            Object.keys(params).forEach((key) => {
                path += `/${params[key]}`;
            });
        }

        this.router.show(path);
    }

    /**
     * Subscribe to route changes.
     *
     * @param {Object} observer - An observer that will be notified on route change.
     *                      Observers are objects with a next() method that will be invoked
     *                      with the new route every time the route changes.
     * @returns A subscription object with an unsubscribe() method that can be used to stop
     *          receiving notifications.
     * @example
     *  const subscription = subscribe({
     *      next: (route) => {
     *          // handle route change
     *      }
     *  });
     *
     *  // later
     *  subscription.unsubscribe();
     *
     */
    subscribe(observer) {
        if (this.currentRoute) {
            observer.next(this.currentRoute);
        }

        this.observers.add(observer);

        return {
            unsubscribe: () => {
                this.observers.delete(observer);
            }
        };
    }

    /**
     * Sets the 3rd party router, currently https://visionmedia.github.io/page.js/.
     * We do this so that we can compile in core where the 3rd party router is not available.
     *
     * Also resets current routes and observer since they are tied to the current router
     * although setting router impl multiple times should only happen in tests.
     *
     * @param {Object} routerImpl - 3rd party router
     */
    setRouter(routerImpl) {
        this.router = routerImpl;
    }
}

// create an instance with bound methods so that they can be exported
const instance = autoBind(new RoutingService());

export const { registerRoutes, getRouteUrl, getRoute, navigateToRoute, subscribe, setRouter } = instance;

export default { registerRoutes, getRouteUrl, getRoute, navigateToRoute, subscribe, setRouter };