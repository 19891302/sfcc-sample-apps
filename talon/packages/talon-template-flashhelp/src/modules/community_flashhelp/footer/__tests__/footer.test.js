import { createElement } from 'lwc';
import { registerRoutes, setRouter } from 'talon/routingService';
import { register as registerConfigProvider } from 'talon/configProvider';
import footer from 'community_flashhelp/footer';
import routes from '../../../../routes.json';

function router() {}
router.base = () => {};
router.start = () => {};

setRouter(router);

registerRoutes(routes);

registerConfigProvider({
    getBasePath() {
        return "/base";
    }
});

describe('community_flashhelp/footer', () => {
    beforeEach(() => {
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    it('renders correctly', () => {
        const element = createElement('community_flashhelp-footer', { is: footer });
        document.body.appendChild(element);

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });
});
