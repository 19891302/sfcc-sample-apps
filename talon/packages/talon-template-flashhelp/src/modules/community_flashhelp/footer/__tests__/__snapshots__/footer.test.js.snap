// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`community_flashhelp/footer renders correctly 1`] = `
<community_flashhelp-footer>
  #shadow-root(open)
    <footer
      class="footer"
    >
      <div
        class="container-fluid slds-text-align_center"
      >
        <p>
          <span>
            <small>
              Â© 2018 The American National Red Cross
            </small>
          </span>
        </p>
        <span>
          <small>
            For more information, please visit redcross.org or cruzrojaamericana.org
          </small>
        </span>
        <div
          class="slds-align_absolute-center"
        >
          <ul
            class="slds-list_horizontal slds-has-inline-block-links_space"
          >
            <li>
              <talon-router-link>
                #shadow-root(open)
                  <a
                    href="/base/"
                  >
                    Home
                  </a>
              </talon-router-link>
            </li>
            <li>
              <talon-router-link>
                #shadow-root(open)
                  <a
                    href="/base/about"
                  >
                    About
                  </a>
              </talon-router-link>
            </li>
            <li>
              <talon-router-link>
                #shadow-root(open)
                  <a
                    href="/base/faq"
                  >
                    FAQ
                  </a>
              </talon-router-link>
            </li>
          </ul>
        </div>
      </div>
    </footer>
</community_flashhelp-footer>
`;
